pkg/services/ngalert/state/manager.go

```diff
@@ -177,26 +177,22 @@ func (st *Manager) ProcessEvalResults(ctx context.Context, evaluatedAt time.Time
 		s := st.setNextState(ctx, alertRule, result, extraLabels, logger)
 		states = append(states, s)
 	}
-	resolvedStates := st.staleResultsHandler(ctx, evaluatedAt, alertRule, logger)
+	staleStates := st.deleteStaleStatesFromCache(ctx, logger, evaluatedAt, alertRule)
+	st.deleteAlertStates(ctx, logger, staleStates)
```

Comment:
Do we need to split this into two in order to unify the history calls? As far as I can tell, `deleteAlertStates` is the only consumer of `staleStates` that cares about non-resolved states. Every time `staleStates` is read beyond this point, we have to do extra work to filter out all the states where `PreviousState != eval.Alerting`, which seems needlessly complicated.

If we keep them together and only return resolved states, we save a pass and greatly simplify the deletion code. We can still delay the saving of state transitions and condense it into a single call without needing to re-work this:
```
resolvedStates := st.deleteStaleStates(...)  // deletes from both cache and database like before, does not call historian

st.saveAlertStates(ctx, logger, states...)

states = append(states, resolvedStates...)

logStateTransitions(ctx, rule, states)

nextStates := make([]*State, 0, len(states))
for _, s := range staleStates {
    nextStates = append(nextStates, s.State)
}
```
There is still only exactly one call to `historian` and one call to `instanceStore` in the above.

Reply:
this is not just about writing state history in one place but also about consolidating all db operations in one place. This refactoring highlights the difference in handling regular states and stale states, and I think we should change the logic of filtering stale states so we can put all states together and then handle them the same way.

Reply:
> put all states together and then handle them the same way

I agree, that's what I was trying to do here by combining resolved states with the newly evaluated states as early as possible.

Reply:
I would prefer to not change the logic in this PR. It's only a refactoring.